.adr/0007-virtualdefund-protocol.md

```diff
@@ -0,0 +1,69 @@
+
+## Decision
+
+- Alice instigates the virtualdefund protocol.
+- She sends an `{isFinal: true, turnNum: 3}` state for `V` "adjusted" by the latest (largest amount) voucher to Bob. Adjusted means the voucher amount is deducted from Alice's allocation and added to Bob's.
+- Bob verifies that state, requiring the adjustment amount to be equal to or larger than the largest voucher he has received.
+- Bob forwards Alice's signature and his own signature on the final state to all intermediaries.
+- The intermediaries countersign and broadcast their signature to all other participants.
+- The leader in each ledger channel proposes to defund `V`.
+- The follower in each ledger channel agrees (countersigns the ledger update).
+
+## Justification
+
+Although there may be some benefit to having a “symmetric” protocol — we wouldn’t have to switch on the participant role as much when implementing it -- we are instead adopting an asymmetric protocol with each role behaving rather differently.
+
+The roles in the virtual channel are very distinct. We can leverage these distinctions to end up with a simple protocol:
+
+1. There is a strong incentive bias between Alice and Bob (at least for a simplex channel).
+2. Intermediaries (e.g. Irene) are not involved in voucher exchange at all during payments (in fact, this is the main selling point of virtual channels), so they need not be involved with vouchers during defunding either. As long as there is a single voucher amount agreed by Alice and Bob (i.e. resolved into the final state of `V`), their position remains neutral independent of whatever that amount is.
+
+### Details
+
+Alice will be programmed to construct a final state for V **which has its outcome adjusted with the latest voucher.** There is an opportunity for her to be malicious here and send a lower voucher / stale voucher. She sends the state to Bob in a message. The message has a header (or type) “virtual defund instigation message” and includes sufficient data for Bob to verify that the update is acceptable: the final state and the signature. He will need to compute the effective voucher amount himself (Alice could send it, but Bob needs to verify anyway).
+
+Bob’s client will get this message and trigger a verification procedure. All he cares about is that the adjusted outcome is **as good or better as** he expects given _his_ latest voucher. So if the special message comes in ahead of the actual latest voucher, he will be pleasantly surprised and continue with the protocol. If a malicious message comes in (Alice knowingly sending a low voucher), he will ignore it.
+
+At this point Alice and Bob have agreed on a final state with an appropriate adjustment. Irene did not need to be involved in that. Bob sends his signature and Alice’s signature to Irene. When she has both signatures, she adds her own and broadcasts to Alice and Bob.
```

Comment:
Could Irene optimistically sign the final state before Bob validates the latest amount? It seems more efficient to have Alice broadcast the signed final state to both Bob and Irene at the same time. Irene can blindly sign it and then wait on Bob to validate and sign it as well.

If Alice is malicious I don't think it matters if Irene signs the state,  as we can rely on Bob to verify the state and only sign it if the amount is big enough.

Reply:
Interesting question. This got me thinking about what the purpose of Irene's signature on the final state is. 

Irene cares about the following things:
1. She must be able to force the channel to close, since here money "points at it" in two ledger channels, and she may need to `reclaim` it.
2. The channel must close with a **reclaimable** outcome. That is, one which allocates the same total funds to the same destinations as the postfund state. if there is no money allocated, no money can be `reclaimed`. If the expected destinations aren't in the outcome, `reclaim` will revert:
https://github.com/statechannels/go-nitro/blob/c5e051a95ded01ce51017809bd84a5192391259f/nitro-protocol/contracts/MultiAssetHolder.sol#L377
3. When the channel closes, she must be certain about which outcome will "win" on chain. Otherwise, she might update one ledger channel based on a voucher of `x`, and then see the other ledger channel challenged as well as `V` challenged with a voucher of `y`. She can lose money this way. Once she gets a full set of signatures on a final state, she has this certainty since final states trump all other states. That's true both when challenging:
https://github.com/statechannels/go-nitro/blob/c5e051a95ded01ce51017809bd84a5192391259f/nitro-protocol/contracts/VirtualPaymentApp.sol#L37-L40
(since the final turnnum is greatest), and when shortcutting with `conclude`:
https://github.com/statechannels/go-nitro/blob/c5e051a95ded01ce51017809bd84a5192391259f/nitro-protocol/contracts/ForceMove.sol#L146
Therefore she can happily proceed to defund off chain with voucher `x`, since she knows she can force this on chain. 

So I think Irene definitely needs to check for reclaim-ability before signing `V.isFinal`. And one she has signed `V.isFinal`, she should never sign it again!

I think none of this prevents your idea of optimistic signing. I've also spotted that Alice never receives Bob's `V.isFinal` signature. I'll update the ADR.

Reply:
[e0a336b](https://github.com/statechannels/go-nitro/pull/863/commits/e0a336bf031d450edf4d7f6f010df8b11fbb61fc)

Reply:
We save an entire round of network comms using your suggestion.